{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "#Copyright © 2024 LOCBP @ University of Zürich\n",
        "#Distributed under MIT license\n",
        "\n",
        "APPEND_JOB = False \n",
        "\"\"\"\n",
        "INPUT PARAMETERS\n",
        "\"\"\"\n",
        "#Do not use the words \"design\", \"rank\" in the names or pdb files as these ar keywords for functionality of the code\n",
        "\n",
        "\"\"\"RFdiffusion\"\"\"\n",
        "NAME = \"\" #if empty, the name will be the PDB #No whitespaces\n",
        "PDB = \"\" #pdb (upload not implemented)\n",
        "CONTIGS = \"\" #[] will be added automatically \n",
        "INPAINT = \"\" #Same format as contigs, it determines which aminoacids of the original pdb to mask\n",
        "NUM_DESIGNS = 25\n",
        "\"\"\"\n",
        "In this notebook, a first RFdiffusion scan is made, with number of designs equal to INPAINT_AUTO_NUM_DESIGNS.\n",
        "Then, no ProteinMPNN/Folding are performed.\n",
        "Rather, the generated structures are used to determine which aminoacids of the fixed chains tend to be close to the ones that are designed (Distance <= INPAINT_AUTO_DISTANCE).\n",
        "If an aminoacid happens to be close to the desined ones at least a INPAINT_AUTO_MIN_OCCURENCY number of times, that it is chosen as be inpainted, unless it is part of the selection to be excluded.\n",
        "The resulting list of aminoacids to be inpainted is combined with the one in INPAINT, and then the following is as in the basic RFdiffusion notebook.\n",
        "\"\"\"\n",
        "INPAINT_AUTO_NUM_DESIGNS = 10\n",
        "INPAINT_AUTO_MIN_OCCURENCY = INPAINT_AUTO_NUM_DESIGNS/5\n",
        "INPAINT_AUTO_DISTANCE = 5 #Å\n",
        "INPAINT_AUTO_EXCLUDE = \"\" #Pymol formatted selection\n",
        "#Advanced\n",
        "ACTIVE_SITE = False #True if you're scaffolding a very small motif\n",
        "STEPS = 50 #Don't recommend to modify this\n",
        "PARTIAL_STEPS = 0 #Steps of partial diffusion. Contig string must be exactly the same length as the input protein (see official documentation)\n",
        "REPRODUCIBLE = False\n",
        "REPRODUCIBILITY_NUMBER = 0\n",
        "\"\"\"ProteinMPNN\"\"\"\n",
        "NUM_SEQ_PER_TARGET = 2\n",
        "FIXED = \"rfd\" #Can be either \"\" (none) or \"rfd\" (scaffold is fixed) or a pymol-formatted selection e.g. \"10-15+17+20-54\" (Only with fixed contigs)\n",
        "FIXED_CHAIN = \"A\" #To be implemented: possibility to fix more than one chain\n",
        "#Advanced\n",
        "SAMPLING_T = \"0.1\"\n",
        "\"\"\"FOLD\"\"\"\n",
        "FOLD = \"OF\" #Choices are AF for AlphaFold and OF for OmegaFold\n",
        "#AlphaFold\n",
        "ONLY_FIRST = True #Only compare the best folding of each sequence generated \n",
        "#Advanced\n",
        "NUM_RELAX = 0 #How many of the best-ranked models do you want to relax with amber?\n",
        "NUM_RECYCLE = 3 #Default (and recommended) is 3\n",
        "RAND_SEED = 0\n",
        "#OmegaFold\n",
        "MODEL = 2 #Model 2 can only be used with V100-32GB or A100-80GB\n",
        "#Ranking\n",
        "METRIC = \"pLDDT\" #\"pLDDT\" or \"pTM\" or \"RMSD\" (Fold) or \"score\" or \"global_score\" (PMPNN). pTM not available in omegafold\n",
        "RMSD_ALIGNMENT = 'super' #'cealign' or 'align' or 'super' or 'fit'\n",
        "PYMOL_BEST = min(max(5,int(NUM_DESIGNS/5)),20) #Create a pymol session contaning the N best designs\n",
        "#Refolding and reranking for models other than alphafold\n",
        "FOLD_BEST_WITH_ALPHAFOLD = PYMOL_BEST \n",
        "#DNA\n",
        "DNA=True #Generates DNA sequences for Homo Sapiens and E coli\n",
        "\n",
        "\"\"\"\n",
        "CODE\n",
        "\"\"\"\n",
        "#Quick refinement of input\n",
        "ONE_JOB = not APPEND_JOB\n",
        "if FOLD == \"AF\": FOLD = \"AlphaFold\"\n",
        "elif FOLD == \"OF\": FOLD = \"OmegaFold\"\n",
        "if INPAINT == \"\": INPAINT=\"-\"\n",
        "if INPAINT_AUTO_EXCLUDE == \"\": INPAINT_AUTO_EXCLUDE=\"-\"\n",
        "INPAINT_AUTO_MIN_OCCURENCY = int(INPAINT_AUTO_MIN_OCCURENCY)\n",
        "\n",
        "import os,shutil #WE USE ABSOLUTE PATHS\n",
        "#CUDA\n",
        "ENVIRONMENT = \"ProteinEnv\"\n",
        "#General\n",
        "MODELS_OUTPUT_FOLDER = \"scratch\" #Where to store output (subfolders will be created inside)\n",
        "INSTALLATION_FOLDER = \"data\"\n",
        "#GENERAL FOLDERS\n",
        "NOTEBOOKS_FOLDER = os.getcwd()\n",
        "PIPELINES_FOLDER = os.path.join(NOTEBOOKS_FOLDER,\"Pipelines\")\n",
        "PDBs_FOLDER = os.path.join(NOTEBOOKS_FOLDER,\"PDBs\")\n",
        "HELP_FOLDER = os.path.join(NOTEBOOKS_FOLDER,\"HelpScripts\")\n",
        "USER_NAME = os.path.basename(os.path.dirname(NOTEBOOKS_FOLDER))\n",
        "HOME_FOLDER = f\"/home/{USER_NAME}\"\n",
        "DATA_FOLDER = f\"/data/{USER_NAME}\"\n",
        "SCRATCH_FOLDER = f\"/scratch/{USER_NAME}\"\n",
        "#MODIFIABLE FOLDERS\n",
        "INSTALLATION_FOLDER = DATA_FOLDER\n",
        "MODELS_OUTPUT_FOLDER = os.path.join(SCRATCH_FOLDER,\"ProteinOutput\")\n",
        "if not os.path.exists(MODELS_OUTPUT_FOLDER):\n",
        "    os.mkdir(MODELS_OUTPUT_FOLDER)\n",
        "#MODELS\n",
        "RFD_FOLDER = os.path.join(INSTALLATION_FOLDER,\"RFdiffusion\")\n",
        "PMPNN_FOLDER = os.path.join(INSTALLATION_FOLDER,\"ProteinMPNN\")\n",
        "OMEGA_FOLDER = os.path.join(INSTALLATION_FOLDER,\"OmegaFold\")\n",
        "COLAB_FOLD_FOLDER = os.path.join(INSTALLATION_FOLDER,\"localcolabfold\")\n",
        "OUT_FOLDER = os.path.join(MODELS_OUTPUT_FOLDER,\"RFdiffusion_InPaint\")\n",
        "if not os.path.exists(OUT_FOLDER):\n",
        "    os.mkdir(OUT_FOLDER)\n",
        "\n",
        "#This will be used throughout to generate file and directory names to avoid overriding old outputs\n",
        "def unique_name(directory,root,ext = \"\",fullpath=0,w=3):   \n",
        "    i = 1     \n",
        "    u_name = root + \"_\" + \"{:0>{width}}\".format(i, width=w)\n",
        "    while os.path.exists(os.path.join(directory,u_name+ext)):\n",
        "        i += 1\n",
        "        u_name = root + \"_\" + \"{:0>{width}}\".format(i, width=w)\n",
        "    if fullpath: return os.path.join(directory, u_name + ext)\n",
        "    return u_name + ext\n",
        "\n",
        "\"\"\"\n",
        "SET JOB NAME AND CREATE OUTPUT DIRECTORIES\n",
        "Data are stored in the \"RFD_output\" folder inside the data folder\n",
        "Also, a folder containg all the results is created inside that folder\n",
        "\"\"\"\n",
        "JOB_BASE = NAME if NAME != \"\" else os.path.splitext(os.path.basename(PDB))[0]\n",
        "JOB = unique_name(OUT_FOLDER,JOB_BASE)\n",
        "JOB_FOLDER = os.path.join(OUT_FOLDER,JOB)\n",
        "os.mkdir(JOB_FOLDER)\n",
        "RUNTIME_FOLDER = os.path.join(JOB_FOLDER,\"RunTime\")\n",
        "if not os.path.exists(RUNTIME_FOLDER):\n",
        "    os.mkdir(RUNTIME_FOLDER)\n",
        "#Output from consol will also be redirected to log files\n",
        "pre_rfd_log_file = os.path.join(JOB_FOLDER,\"_pre_rfd.log\")\n",
        "inpaint_log_file = os.path.join(JOB_FOLDER,\"_inpaint.log\")\n",
        "rfd_log_file = os.path.join(JOB_FOLDER,\"_rfd.log\")\n",
        "pmpnn_log_file = os.path.join(JOB_FOLDER,\"_pmpnn.log\")\n",
        "fold_log_file = os.path.join(JOB_FOLDER,\"_fold.log\")   #####changed af2 to fold\n",
        "ranking_log_file = os.path.join(JOB_FOLDER,\"_ranking.log\")\n",
        "dna_log_file = os.path.join(JOB_FOLDER,\"_dna.log\")\n",
        "\n",
        "\"\"\"\n",
        "PRELIMINARY RFdiffusion to determine inpainted positions\n",
        "\"\"\"\n",
        "PRE_RFD_JOB_FOLDER = os.path.join(JOB_FOLDER,\"RFdiffusion-InPaint\")\n",
        "pre_rfd_options = f\"'contigmap.contigs=[{CONTIGS}]'\"\n",
        "if INPAINT != \"-\": pre_rfd_options += f\" 'contigmap.inpaint_seq=[{INPAINT}]'\"\n",
        "pdb_file = PDB\n",
        "if PDB != \"\": \n",
        "    pdb_temp = PDB if PDB.endswith(\".pdb\") else PDB + \".pdb\"\n",
        "    pdb_temp_file = os.path.join(PDBs_FOLDER,pdb_temp)\n",
        "    pdb_file = os.path.join(RUNTIME_FOLDER,pdb_temp)\n",
        "    shutil.copy(pdb_temp_file,pdb_file) #copy original input file into destination foler\n",
        "    pre_rfd_options += f\" inference.input_pdb={pdb_file}\"\n",
        "prefix = os.path.join(PRE_RFD_JOB_FOLDER,f\"{JOB}_pre_design\") #polyG Designs will be .._design_0.pdb, .._design_1.pdb, ...\n",
        "pre_rfd_options += f\" inference.output_prefix={prefix}\"\n",
        "pre_rfd_options += f\" inference.num_designs={INPAINT_AUTO_NUM_DESIGNS}\"\n",
        "pre_rfd_options += f\" inference.deterministic={REPRODUCIBLE}\"\n",
        "pre_rfd_options += f\" inference.design_startnum={REPRODUCIBILITY_NUMBER}\"\n",
        "if STEPS != 50: pre_rfd_options += f\" diffuser.T = {STEPS}\"\n",
        "if PARTIAL_STEPS > 0: pre_rfd_options += f\" diffuser.partial_T = {PARTIAL_STEPS}\"\n",
        "if ACTIVE_SITE: pre_rfd_options += \" inference.ckpt_override_path=models/ActiveSite_ckpt.pt\"\n",
        "\n",
        "inference_py_file = os.path.join(RFD_FOLDER,\"scripts\",\"run_inference.py\")\n",
        "pre_rfd_cmd = f\"\"\"\n",
        "{inference_py_file} {pre_rfd_options}\n",
        "\"\"\"\n",
        "pre_rfd_sh_file = os.path.join(RUNTIME_FOLDER,\"pre_rfd.sh\")\n",
        "with open(pre_rfd_sh_file ,\"w\") as pre_rfd_sh:\n",
        "    pre_rfd_sh.write(pre_rfd_cmd)\n",
        "os.chmod(pre_rfd_sh_file, 0o755) #By default, the bash is created without execution rights, we add them here\n",
        "\n",
        "\n",
        "\"\"\"\n",
        "RFDiffusion\n",
        "\"\"\"\n",
        "RFD_JOB_FOLDER = os.path.join(JOB_FOLDER,\"RFdiffusion\")\n",
        "rfd_options = f\"'contigmap.contigs=[{CONTIGS}]'\"\n",
        "#####Inpaint will be handled afterwards\n",
        "pdb_file = PDB\n",
        "if PDB != \"\": \n",
        "    pdb_temp = PDB if PDB.endswith(\".pdb\") else PDB + \".pdb\"\n",
        "    pdb_temp_file = os.path.join(PDBs_FOLDER,pdb_temp)\n",
        "    pdb_file = os.path.join(RUNTIME_FOLDER,pdb_temp)\n",
        "    shutil.copy(pdb_temp_file,pdb_file) #copy original input file into destination foler\n",
        "    rfd_options += f\" inference.input_pdb={pdb_file}\"\n",
        "prefix = os.path.join(RFD_JOB_FOLDER,f\"{JOB}_design\") #polyG Designs will be .._design_0.pdb, .._design_1.pdb, ...\n",
        "rfd_options += f\" inference.output_prefix={prefix}\"\n",
        "rfd_options += f\" inference.num_designs={NUM_DESIGNS}\"\n",
        "rfd_options += f\" inference.deterministic={REPRODUCIBLE}\"\n",
        "rfd_options += f\" inference.design_startnum={REPRODUCIBILITY_NUMBER}\"\n",
        "if STEPS != 50: rfd_options += f\" diffuser.T = {STEPS}\"\n",
        "if PARTIAL_STEPS > 0: rfd_options += f\" diffuser.partial_T = {PARTIAL_STEPS}\"\n",
        "if ACTIVE_SITE: rfd_options += \" inference.ckpt_override_path=models/ActiveSite_ckpt.pt\"\n",
        "\n",
        "inference_py_file = os.path.join(RFD_FOLDER,\"scripts\",\"run_inference.py\")\n",
        "rfd_cmd = f\"\"\"\n",
        "{inference_py_file} {rfd_options}\n",
        "\"\"\"\n",
        "rfd_sh_file = os.path.join(RUNTIME_FOLDER,\"rfd.sh\")\n",
        "with open(rfd_sh_file ,\"w\") as rfd_sh:\n",
        "    rfd_sh.write(rfd_cmd)\n",
        "os.chmod(rfd_sh_file, 0o755) #By default, the bash is created without execution rights, we add them here\n",
        "\n",
        "\"\"\"\n",
        "Determine position and determine which positions to inpaint\n",
        "Then will simply open the file rfd_sh_file and add the flag for the inpaint\n",
        "\"\"\"\n",
        "inpaint_py_file = os.path.join(HELP_FOLDER,\"find_inpaint.py\")\n",
        "inpaint_csv_file = os.path.join(JOB_FOLDER,JOB_BASE+\"_inpaint.csv\")\n",
        "inpaint_pse_file = os.path.join(JOB_FOLDER,JOB_BASE+\"_inpaint.pse\")\n",
        "inpaint_sh_file = os.path.join(RUNTIME_FOLDER,\"find_inpaint.sh\")\n",
        "inpaint_cmd = f\"\"\"\n",
        "python {inpaint_py_file} {PRE_RFD_JOB_FOLDER} {pre_rfd_log_file} {rfd_sh_file} {CONTIGS} {INPAINT} {INPAINT_AUTO_NUM_DESIGNS} {INPAINT_AUTO_DISTANCE} {INPAINT_AUTO_MIN_OCCURENCY} {INPAINT_AUTO_EXCLUDE} {pdb_file} {inpaint_csv_file} {inpaint_pse_file}\n",
        "\"\"\"\n",
        "with open(inpaint_sh_file,'w') as inpaint_sh:\n",
        "    inpaint_sh.write(inpaint_cmd)\n",
        "os.chmod(inpaint_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "\"\"\"ProteinMPNN\"\"\"\n",
        "#Parse folder for multiple pdb input\n",
        "parsed_pdbs_jsonl_file = os.path.join(RUNTIME_FOLDER,f\"{JOB}_parsed_pdbs.jsonl\")\n",
        "parse_py_file = os.path.join(PMPNN_FOLDER,\"helper_scripts/parse_multiple_chains.py\")\n",
        "fixed_jsonl_file = os.path.join(RUNTIME_FOLDER,f\"{JOB}_fixed_pos.jsonl\")\n",
        "sele_csv_file = os.path.join(RUNTIME_FOLDER,\"pmpnn_sele.csv\")\n",
        "\n",
        "#IMPLEMENTATION OF FIXED POSITIONS\n",
        "fixed_py_file = os.path.join(HELP_FOLDER,\"make_fixed_dict.py\")\n",
        "\n",
        "#Set options\n",
        "pmpnn_options = f\"--num_seq_per_target {NUM_SEQ_PER_TARGET}\"\n",
        "pmpnn_options += f\" --sampling_temp {SAMPLING_T}\"\n",
        "pmpnn_py_file = os.path.join(PMPNN_FOLDER,\"protein_mpnn_run.py\")\n",
        "\n",
        "pmpnn_sh_file = os.path.join(RUNTIME_FOLDER,\"rfd_pmpnn.sh\")\n",
        "pmpnn_cmd = f\"\"\"\n",
        "echo Determining fixed positions\n",
        "python {fixed_py_file} {RFD_JOB_FOLDER} {rfd_log_file} 100 {FIXED} {FIXED_CHAIN} {fixed_jsonl_file} {sele_csv_file}\n",
        "echo Parsing multiple pbds \n",
        "python {parse_py_file} --input_path {RFD_JOB_FOLDER} --output_path {parsed_pdbs_jsonl_file}\n",
        "echo Running model\n",
        "python {pmpnn_py_file} --jsonl_path {parsed_pdbs_jsonl_file} --fixed_positions_jsonl {fixed_jsonl_file} --out_folder {JOB_FOLDER} {pmpnn_options}\n",
        "\"\"\"\n",
        "with open(pmpnn_sh_file,'w') as pmpnn_sh:\n",
        "    pmpnn_sh.write(pmpnn_cmd)\n",
        "os.chmod(pmpnn_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "\"\"\"FOLD\"\"\"\n",
        "FOLD_OUT_FOLDER = os.path.join(JOB_FOLDER,FOLD)\n",
        "if not os.path.exists(FOLD_OUT_FOLDER):\n",
        "    os.mkdir(FOLD_OUT_FOLDER)\n",
        "PMPNN_FA_FOLDER = os.path.join(JOB_FOLDER,\"seqs\") #This is the folder where pmpnn outputs fasta files\n",
        "queries_csv_file = os.path.join(RUNTIME_FOLDER,JOB_BASE+\"_queries.csv\") #These will contain the queries\n",
        "queries_fasta_file = os.path.join(JOB_FOLDER,JOB_BASE+\"_queries.fasta\")\n",
        "fa_to_csv_fasta_py_file = os.path.join(HELP_FOLDER,\"fa_to_csv_fasta.py\")\n",
        "\n",
        "fold_sh_file = \"\"\n",
        "if FOLD == \"AlphaFold\":\n",
        "    af2_options = \"\"\n",
        "    if NUM_RELAX > 0: af2_options += f\" --amber --use-gpu-relax --num-relax {NUM_RELAX}\" #assumed to run on GPU\n",
        "    if NUM_RECYCLE != 3: af2_options += f\" --num-recycle {NUM_RECYCLE}\"\n",
        "    if RAND_SEED != 0: af2_options += f\" --random-seed {RAND_SEED}\"\n",
        "\n",
        "    colabfold_batch_file = os.path.join(COLAB_FOLD_FOLDER,\"colabfold-conda/bin/colabfold_batch\")\n",
        "    fold_sh_file = os.path.join(RUNTIME_FOLDER,\"rfd_pmpnn_af2.sh\")\n",
        "    af2_cmd = f\"\"\"\n",
        "    echo Generating queries.csv file from .fa output\n",
        "    python {fa_to_csv_fasta_py_file} {PMPNN_FA_FOLDER} {queries_csv_file} {queries_fasta_file}\n",
        "    echo Initializing model...\n",
        "    {colabfold_batch_file} {queries_csv_file} {FOLD_OUT_FOLDER} {af2_options}\n",
        "    \"\"\"\n",
        "    with open(fold_sh_file,'w') as af2_sh:\n",
        "        af2_sh.write(af2_cmd)\n",
        "elif FOLD == \"OmegaFold\":\n",
        "    of_options = \"\"\n",
        "    of_options += f\" --model {MODEL}\"\n",
        "    weights_pt = [\"\",\"release1.pt\",\"release2.pt\"][MODEL]\n",
        "    weights_pt_file = os.path.join(OMEGA_FOLDER,weights_pt)\n",
        "    of_options += f\" --weights_file {weights_pt_file}\"\n",
        "    fold_sh_file = os.path.join(RUNTIME_FOLDER,\"rfd_pmpnn_of.sh\")\n",
        "    of_cmd = f\"\"\"\n",
        "    echo Generating queries.csv file from .fa output\n",
        "    python {fa_to_csv_fasta_py_file} {PMPNN_FA_FOLDER} {queries_csv_file} {queries_fasta_file}\n",
        "    echo Initializing model...\n",
        "    omegafold {queries_fasta_file} {FOLD_OUT_FOLDER} {of_options}\n",
        "    \"\"\"\n",
        "    with open(fold_sh_file,'w') as of_sh:\n",
        "        of_sh.write(of_cmd)\n",
        "os.chmod(fold_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "\"\"\"RANK EVERYTHING\"\"\"\n",
        "rank_pdb = pdb_file if pdb_file != \"\" else \"-\"\n",
        "rank_output_csv_file = os.path.join(JOB_FOLDER,JOB_BASE+\"_\"+FOLD+\".csv\")\n",
        "rank_best_fasta_file = os.path.join(JOB_FOLDER,JOB_BASE+f\"_Best{PYMOL_BEST}.fasta\")\n",
        "rank_pse_file = os.path.join(JOB_FOLDER,JOB_BASE+f\"_{FOLD}.pse\")\n",
        "rank_py_file = os.path.join(HELP_FOLDER,f\"rank_{FOLD}.py\")\n",
        "rank_cmd = f\"python {rank_py_file} {fold_log_file} {queries_csv_file} {NUM_SEQ_PER_TARGET} {sele_csv_file} {FOLD_OUT_FOLDER} {rank_pdb} {rank_output_csv_file} {METRIC} {RMSD_ALIGNMENT} {rank_pse_file} {PYMOL_BEST} {rank_best_fasta_file} {ONLY_FIRST}\"\n",
        "rank_sh_file = os.path.join(RUNTIME_FOLDER,f\"rank_{FOLD}.sh\")\n",
        "with open(rank_sh_file,'w') as rank_sh:\n",
        "    rank_sh.write(rank_cmd)\n",
        "os.chmod(rank_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "\"\"\"DNA\"\"\"\n",
        "DNA_FOLDER=os.path.join(JOB_FOLDER,\"DNA\")\n",
        "if not os.path.exists(DNA_FOLDER):\n",
        "    os.mkdir(DNA_FOLDER)\n",
        "dna_prefix = os.path.join(DNA_FOLDER,JOB_BASE)\n",
        "dna_encoder_py_file = os.path.join(HELP_FOLDER,\"dna_encoder.py\")\n",
        "dna_encoder_cmd = f\"python {dna_encoder_py_file} {rank_best_fasta_file} {dna_prefix}\"\n",
        "dna_encoder_sh_file = os.path.join(RUNTIME_FOLDER,\"dna_encoder.sh\")\n",
        "with open(dna_encoder_sh_file,'w') as dna_encoder_sh:\n",
        "    dna_encoder_sh.write(dna_encoder_cmd)\n",
        "os.chmod(dna_encoder_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "\"\"\"Additional step: fold best ones with AlphaFold\"\"\"\n",
        "if FOLD_BEST_WITH_ALPHAFOLD > 0 and not FOLD==\"AlphaFold\":\n",
        "    refold_log_file = os.path.join(JOB_FOLDER,\"_af2_best_fold.log\")\n",
        "    reranking_log_file = os.path.join(JOB_FOLDER,\"_reranking.log\")\n",
        "\n",
        "    ALPHAFOLD_OUT_FOLDER = os.path.join(JOB_FOLDER,\"AlphaFold\")\n",
        "    if not os.path.exists(ALPHAFOLD_OUT_FOLDER):\n",
        "        os.mkdir(ALPHAFOLD_OUT_FOLDER)\n",
        "\n",
        "    best_queries_py_file = os.path.join(HELP_FOLDER,\"best_queries.py\")\n",
        "    best_queries_csv_file = os.path.join(RUNTIME_FOLDER,\"best_queries.csv\")\n",
        "\n",
        "    af2_options = \"\"\n",
        "    if NUM_RELAX > 0: af2_options += f\" --amber --use-gpu-relax --num-relax {NUM_RELAX}\" #assumed to run on GPU\n",
        "    if NUM_RECYCLE != 3: af2_options += f\" --num-recycle {NUM_RECYCLE}\"\n",
        "    if RAND_SEED != 0: af2_options += f\" --random-seed {RAND_SEED}\"\n",
        "\n",
        "    colabfold_batch_file = os.path.join(COLAB_FOLD_FOLDER,\"colabfold-conda/bin/colabfold_batch\")\n",
        "    af2_refolding_sh_file = os.path.join(RUNTIME_FOLDER,\"rfd_pmpnn_refold_af2.sh\")\n",
        "    af2_refolding_cmd = f\"\"\"\n",
        "    echo Selecting best queries {FOLD} output\n",
        "    python {best_queries_py_file} {queries_csv_file} {rank_output_csv_file} {FOLD_BEST_WITH_ALPHAFOLD} {best_queries_csv_file}\n",
        "    echo Initializing model...\n",
        "    {colabfold_batch_file} {best_queries_csv_file} {ALPHAFOLD_OUT_FOLDER} {af2_options}\n",
        "    \"\"\"\n",
        "    with open(af2_refolding_sh_file,'w') as af2_sh:\n",
        "        af2_sh.write(af2_refolding_cmd)\n",
        "    os.chmod(af2_refolding_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "    \"\"\"RERANK EVERYTHING\"\"\"\n",
        "    rerank_output_csv_file = os.path.join(JOB_FOLDER,JOB_BASE+\"_AlphaFold.csv\")\n",
        "    rerank_pse_file = os.path.join(JOB_FOLDER,JOB_BASE+f\"_AlphaFold.pse\")\n",
        "    rank_py_file = os.path.join(HELP_FOLDER,f\"rank_AlphaFold.py\")\n",
        "    rank_cmd = f\"python {rank_py_file} {refold_log_file} {best_queries_csv_file} {NUM_SEQ_PER_TARGET} {sele_csv_file} {ALPHAFOLD_OUT_FOLDER} {rank_pdb} {rerank_output_csv_file} {METRIC} {RMSD_ALIGNMENT} {rerank_pse_file} {PYMOL_BEST}  {rank_best_fasta_file} {ONLY_FIRST}\"\n",
        "    rerank_sh_file = os.path.join(RUNTIME_FOLDER,f\"rerank_af2.sh\")\n",
        "    with open(rerank_sh_file,'w') as rank_sh:\n",
        "        rank_sh.write(rank_cmd)\n",
        "    os.chmod(rerank_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "\"\"\"COMBINE ALL IN A UNIQUE PIPELINE\"\"\"\n",
        "pipeline_sh_file = unique_name(RUNTIME_FOLDER,\"RFdiffusion_InPaint_pipeline\",\".sh\",1)\n",
        "print_options_cmd = f\"\"\"\n",
        "echo Pipeline: {os.path.basename(pipeline_sh_file)}\n",
        "echo\n",
        "echo \"Name: {NAME}\"\n",
        "echo \"PDB: {PDB}\"\n",
        "echo \"PDB PATH: {pdb_file}\"\n",
        "echo\n",
        "echo \"Determination of inpainted positions\"\n",
        "echo \"  NUM DESIGNS: {INPAINT_AUTO_NUM_DESIGNS}\"\n",
        "echo \"  MAX DISTANCE: {INPAINT_AUTO_DISTANCE}\"\n",
        "echo \"  MIN OCCURANCY: {INPAINT_AUTO_MIN_OCCURENCY}\"\n",
        "echo \"  EXCLUDE: {INPAINT_AUTO_EXCLUDE}\"\n",
        "echo\n",
        "echo \"RFdiffusion\"\n",
        "echo \"  CONTIGS: {CONTIGS}\"\n",
        "echo \"  NUM DESIGNS: {NUM_DESIGNS}\"\n",
        "echo \"  ACTIVE SITE: {ACTIVE_SITE}\"\n",
        "echo \"  STEPS: {STEPS}\"\n",
        "echo \"  PARTIAL STEPS: {PARTIAL_STEPS}\"\n",
        "echo\n",
        "echo \"ProteinMPNN\"\n",
        "echo \"  NUM SEQUENCES PER TARGET: {NUM_SEQ_PER_TARGET}\"\n",
        "echo \"  FIXED: {FIXED}\"\n",
        "echo \"  FIXED CHAIN: {FIXED_CHAIN}\"\n",
        "echo \"  SAMPLING T: {SAMPLING_T}\"\n",
        "echo\n",
        "echo \"FOLD MODEL: {FOLD}\"\n",
        "echo \"  AlphaFold\"\n",
        "echo \"    NUM RELAX: {NUM_RELAX}\"\n",
        "echo \"    NUM RECYCLE: {NUM_RECYCLE}\"\n",
        "echo \"    RANDOM GENERATOR SEED: {RAND_SEED}\"\n",
        "echo \"  OmegaFold\"\n",
        "echo \"    MODEL: {MODEL}\"\n",
        "echo\n",
        "echo \"Alignment algorithm: {RMSD_ALIGNMENT}\"\n",
        "\"\"\"\n",
        "pipeline_cmd = print_options_cmd + f\"\"\"\n",
        "source activate {ENVIRONMENT}\n",
        "echo\n",
        "echo Preliminary RFdiffusion\n",
        "echo\n",
        "{pre_rfd_sh_file} | tee {pre_rfd_log_file}\n",
        "echo Inpainting determination\n",
        "echo\n",
        "{inpaint_sh_file} | tee {inpaint_log_file}\n",
        "echo\n",
        "echo RFDiffusion\n",
        "{rfd_sh_file} | tee {rfd_log_file}\n",
        "echo\n",
        "echo ProteinMPNN\n",
        "{pmpnn_sh_file} | tee {pmpnn_log_file}\n",
        "echo\n",
        "echo {FOLD}\n",
        "{fold_sh_file} | tee {fold_log_file}\n",
        "echo \n",
        "echo Ranking...\n",
        "{rank_sh_file} | tee {ranking_log_file}\n",
        "\"\"\"\n",
        "if DNA: pipeline_cmd += f\"\"\"\n",
        "echo\n",
        "echo DNA\n",
        "{dna_encoder_sh_file} | tee {dna_log_file}\n",
        "\"\"\"\n",
        "if FOLD_BEST_WITH_ALPHAFOLD > 0 and not FOLD==\"AlphaFold\":\n",
        "    pipeline_cmd += f\"\"\"\n",
        "echo\n",
        "echo Refolding best {FOLD_BEST_WITH_ALPHAFOLD} proteins with AlphaFold\n",
        "echo\n",
        "echo AlphaFold\n",
        "{af2_refolding_sh_file} | tee {refold_log_file}\n",
        "echo \n",
        "echo Ranking...\n",
        "{rerank_sh_file} | tee {reranking_log_file}\n",
        "\"\"\"\n",
        "pipeline_cmd += f\"\"\"\n",
        "echo \n",
        "echo Job done\n",
        "\"\"\"\n",
        "zip_py_file = os.path.join(HELP_FOLDER,\"zip_results.py\")\n",
        "pipeline_cmd += f\"\"\"\n",
        "echo\n",
        "echo Zipping results\n",
        "python {zip_py_file} {JOB_FOLDER}\n",
        "echo \n",
        "echo Results in:\n",
        "echo {JOB_FOLDER}\n",
        "\"\"\"\n",
        "with open(pipeline_sh_file,'w') as pipeline_sh:\n",
        "    pipeline_sh.write(pipeline_cmd)\n",
        "os.chmod(pipeline_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "batch_sh_file = os.path.join(PIPELINES_FOLDER,\"rfd_batch.sh\")\n",
        "if not ONE_JOB and os.path.exists(batch_sh_file):\n",
        "    with open(batch_sh_file,\"r\") as rfd_batch_sh:\n",
        "        previous_pipelines = rfd_batch_sh.readlines()\n",
        "    with open(batch_sh_file,\"w\") as rfd_batch_sh:\n",
        "        rfd_batch_sh.writelines(previous_pipelines)\n",
        "        rfd_batch_sh.write(\"\\n\")\n",
        "        rfd_batch_sh.write(pipeline_sh_file)  \n",
        "else:      \n",
        "    with open(batch_sh_file,\"w\") as rfd_batch_sh:\n",
        "        rfd_batch_sh.write(pipeline_sh_file)  \n",
        "os.chmod(batch_sh_file, 0o755) #Give execution rights\n",
        "\n",
        "#write configuration file\n",
        "config_txt_file = os.path.join(JOB_FOLDER,JOB_BASE+\"_config.txt\")\n",
        "with open(config_txt_file,'w') as config_txt:\n",
        "    config_txt.write(print_options_cmd)\n",
        "\n",
        "print(\"Run using next cell\")\n",
        "print(f\"Single job:\\n{pipeline_sh_file}\")\n",
        "print(f\"Batch:\\n{batch_sh_file}\")\n",
        "print(f\"\\nOutput of this pipeline will be in:\\n{JOB_FOLDER}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "%%bash\n",
        "/home/$USER/ProteinNotebooks/Pipelines/rfd_batch.sh"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "\"\"\"\n",
        "MAKE SLURM FILE\n",
        "\"\"\"\n",
        "PACKAGE_MANAGER = \"mamba\"\n",
        "GPU = \"A100\" \n",
        "\"\"\"\n",
        "GPUs available are:\n",
        "t4      0.022 CHF/h\n",
        "v100    0.057 CHF/h\n",
        "a100    0.081 CHF/h\n",
        "These can be allocated by setting GPU = \"t4\", \"v100\", \"v100-32g\" or \"a100\"\n",
        "By setting GPU=\"gpu\", the first available will be used\n",
        "By setting GPU=\"high-memory\", either v100-32gb or a100 are selected. \n",
        "Important: OmegaFold model 2 only works with high-memory GPUs, sometimes crashes even with 32 GB\n",
        "\"\"\"\n",
        "\n",
        "slurm_file = unique_name(PIPELINES_FOLDER,\"rfd_ip_slurm\",\".sh\",1)\n",
        "with open(batch_sh_file,\"r\") as rfd_batch_sh:\n",
        "    all_pipelines = rfd_batch_sh.readlines()\n",
        "with open(slurm_file,\"w\") as slurm_bash:\n",
        "    slurm_bash.write(\"\"\"# Check if nvidia-smi is available\n",
        "if ! command -v nvidia-smi &> /dev/null\n",
        "then\n",
        "    echo \"Could not load GPU correctly: nvidia-smi could not be found\"\n",
        "    exit\n",
        "fi\n",
        "gpu_type=$(nvidia-smi --query-gpu=gpu_name --format=csv,noheader)\n",
        "echo \"GPU Type: $gpu_type\"\\n\n",
        "\"\"\")\n",
        "    slurm_bash.write(f\"module load {PACKAGE_MANAGER}\\n\")\n",
        "    slurm_bash.writelines(all_pipelines)\n",
        "os.chmod(slurm_file, 0o755)\n",
        "\n",
        "if GPU == \"high-memory\":\n",
        "    gpu_line = \"#SBATCH --gpus=1\\n#SBATCH --constraint=\\\"GPUMEM32GB|GPUMEM80GB\\\"\"\n",
        "elif GPU == \"gpu\":\n",
        "    gpu_line = \"#SBATCH --gpus=1\"\n",
        "else:\n",
        "    gpu_line = f\"#SBATCH --gpus={GPU}:1\"\n",
        "\n",
        "job_comps=JOB_FOLDER.split('/')\n",
        "job_name=f\"{job_comps[-2]}: {job_comps[-1]}\"\n",
        "\n",
        "print(f\"\"\"\n",
        "ScienceApps > Jobs > JobComposer > New Job > From Default Template    \n",
        "Edit Job name from Job Options. Suggested:\n",
        "\n",
        "{job_name}       \n",
        "\n",
        "Replace job.sh (Open in Editor) with the following (adapt required time hh:mm:ss) then Save:\n",
        "      \n",
        "#!/usr/bin/bash\n",
        "{gpu_line}\n",
        "#SBATCH --mem=15650\n",
        "#SBATCH --time=23:59:00\n",
        "#SBATCH --output=job.out      \n",
        "{slurm_file}\n",
        "\"\"\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "\"\"\"\n",
        "Conventions for names\n",
        "Paths to files end with _type_file\n",
        "Paths to folders end with _FOLDER\n",
        "Opened files .type end with _type\n",
        "Content of .sh end with _cmd\n",
        "Content of .py end with _script\n",
        "\"\"\"\n",
        "\n",
        "\"\"\"\n",
        "Cleanup instructions\n",
        "Delete data/bash_files\n",
        "Delete outputs folder: contains log files\n",
        "\"\"\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DKQXlWEjIOsf"
      },
      "source": [
        "**Instructions**\n",
        "---\n",
        "---\n",
        "\n",
        "Use `contigs` to define continious chains. Use a `:` to define multiple contigs and a `/` to define mutliple segments within a contig.\n",
        "For example:\n",
        "\n",
        "**unconditional**\n",
        "- `contigs='100'` - diffuse **monomer** of length 100\n",
        "- `contigs='50:100'` - diffuse **hetero-oligomer** of lengths 50 and 100\n",
        "- `contigs='50'` `symmetry='cyclic'` `order=2` - make two copies of the defined contig(s) and add a symmetry constraint, for **homo-oligomeric** diffusion.\n",
        "\n",
        "**binder design**\n",
        "- `contigs='A:50'` `pdb='4N5T'` - diffuse a **binder** of length 50 to chain A of defined PDB.\n",
        "- `contigs='E6-155:70-100'` `pdb='5KQV'` `hotspot='E64,E88,E96'` - diffuse a **binder** of length 70 to 100 (sampled randomly) to chain E and defined hotspot(s).\n",
        "\n",
        "**motif scaffolding**\n",
        " - `contigs='40/A163-181/40'` `pdb='5TPN'`\n",
        " - `contigs='A3-30/36/A33-68'` `pdb='6MRR'` - diffuse a loop of length 36 between two segments of defined PDB ranges.\n",
        "\n",
        "**partial diffusion**\n",
        "- `contigs=''` `pdb='6MRR'` - noise all coordinates\n",
        "- `contigs='A1-10'` `pdb='6MRR'` - keep first 10 positions fixed, noise the rest\n",
        "- `contigs='A'` `pdb='1SSC'` - fix chain A, noise the rest\n",
        "\n",
        "*hints and tips*\n",
        "- `pdb=''` leave blank to get an upload prompt\n",
        "- `contigs='50-100'` use dash to specify a range of lengths to sample from"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "provenance": []
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
